### PR TITLE
csi: update CSIDriverOption params during saving cluster config

### PR DESCRIPTION
**Pull request by Rakshith-R**
During cluster creation, csi config map was first filled with mon ips and without CSIDriverOptions.
This commit makes sure CSIDriverOptions are added at the begining when the entry is first created.

Updates:
- #13835

#### Without fix (kernelmountoptions and readaffinity options are filled in later)
```
[User@localhost ~]:rook$ k get cm rook-ceph-csi-config -o yaml -w
apiVersion: v1
data:
  csi-cluster-config-json: '[]'
kind: ConfigMap
metadata:
  creationTimestamp: "2024-02-29T13:43:36Z"
  name: rook-ceph-csi-config
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: false
    controller: true
    kind: Deployment
    name: rook-ceph-operator
    uid: 17c87c42-f130-481e-996d-a9750b5b21df
  resourceVersion: "10112"
  uid: 7e37de42-06fb-403b-b4a9-f0a8bfc2bb3e
---
apiVersion: v1
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["192.168.39.252:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","kernelMountOptions":"","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNamespace":""},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":null},"namespace":"rook-ceph"}]'
kind: ConfigMap
metadata:
  creationTimestamp: "2024-02-29T13:43:36Z"
  name: rook-ceph-csi-config
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: false
    controller: true
    kind: Deployment
    name: rook-ceph-operator
    uid: 17c87c42-f130-481e-996d-a9750b5b21df
  resourceVersion: "10199"
  uid: 7e37de42-06fb-403b-b4a9-f0a8bfc2bb3e
---
apiVersion: v1
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["192.168.39.252:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","kernelMountOptions":"ms_mode=prefer-crc","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNamespace":""},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":["test"]},"namespace":"rook-ceph"}]'
kind: ConfigMap
metadata:
  creationTimestamp: "2024-02-29T13:43:36Z"
  name: rook-ceph-csi-config
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: false
    controller: true
    kind: Deployment
    name: rook-ceph-operator
    uid: 17c87c42-f130-481e-996d-a9750b5b21df
  resourceVersion: "10306"
  uid: 7e37de42-06fb-403b-b4a9-f0a8bfc2bb3e
```

#### With fix
```
apiVersion: v1
data:
  csi-cluster-config-json: '[]'
kind: ConfigMap
metadata:
  creationTimestamp: "2024-02-29T13:39:13Z"
  name: rook-ceph-csi-config
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: false
    controller: true
    kind: Deployment
    name: rook-ceph-operator
    uid: 74331958-4fc2-492c-b3d7-ccffc9d29e48
  resourceVersion: "124843"
  uid: 8b4f509f-bbb4-4b94-9397-5c75f840e33b
---
apiVersion: v1
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.108.161.177:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","kernelMountOptions":"mds_mode=prefer-crc","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNamespace":""},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":["w"]},"namespace":"rook-ceph"}]'
kind: ConfigMap
metadata:
  creationTimestamp: "2024-02-29T13:39:13Z"
  name: rook-ceph-csi-config
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: false
    controller: true
    kind: Deployment
    name: rook-ceph-operator
    uid: 74331958-4fc2-492c-b3d7-ccffc9d29e48
  resourceVersion: "124946"
  uid: 8b4f509f-bbb4-4b94-9397-5c75f840e33b
```

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by subhamkrai — CHANGES_REQUESTED** [review]
Could you update/add the unit test also?

**Review by travisn — CHANGES_REQUESTED** [review]
A number of integration tests are failing, needs investigation

File: pkg/operator/ceph/csi/cluster_config.go [hunk 1]
```diff
@@ -182,13 +185,21 @@ func updateCsiClusterConfig(curr, clusterKey string, newCsiClusterConfigEntry *C
...
 			}
+			centry.Namespace = clusterKey
 			centry.Monitors = newCsiClusterConfigEntry.Monitors
+			// update subvolumegroup and cephfs netNamespaceFilePath only when either is specified
+			// while always updating kernel and fuse mount options.
 			if newCsiClusterConfigEntry.CephFS.SubvolumeGroup != "" || newCsiClusterConfigEntry.CephFS.NetNamespaceFilePath != "" {
 				centry.CephFS = newCsiClusterConfigEntry.CephFS
+			} else {
+				centry.CephFS.KernelMountOptions = newCsiClusterConfigEntry.CephFS.KernelMountOptions
+				centry.CephFS.FuseMountOptions = newCsiClusterConfigEntry.CephFS.FuseMountOptions
```
Comment (Rakshith-R) [inline pkg/operator/ceph/csi/cluster_config.go:196]: This ensures mount options are always updated.

File: pkg/operator/ceph/csi/cluster_config.go [hunk 2]
```diff
@@ -207,16 +218,9 @@ func updateCsiClusterConfig(curr, clusterKey string, newCsiClusterConfigEntry *C
...
-			// Add a condition not to fill with empty values
-			if newCsiClusterConfigEntry.CephFS.SubvolumeGroup != "" || newCsiClusterConfigEntry.CephFS.NetNamespaceFilePath != "" {
-				centry.CephFS = newCsiClusterConfigEntry.CephFS
-			}
-			if newCsiClusterConfigEntry.NFS.NetNamespaceFilePath != "" {
-				centry.NFS = newCsiClusterConfigEntry.NFS
-			}
+			centry.RBD = newCsiClusterConfigEntry.RBD
+			centry.CephFS = newCsiClusterConfigEntry.CephFS
+			centry.NFS = newCsiClusterConfigEntry.NFS
```
Comment (Rakshith-R) [inline pkg/operator/ceph/csi/cluster_config.go:223]: No need to check non-empty strings while creating a new entry.

**Review by subhamkrai — APPROVED** [review]

File: pkg/operator/ceph/csi/cluster_config.go [hunk 3]
```diff
@@ -183,12 +186,19 @@ func updateCsiClusterConfig(curr, clusterKey string, newCsiClusterConfigEntry *C
 				break
 			}
 			centry.Monitors = newCsiClusterConfigEntry.Monitors
+			// update subvolumegroup and cephfs netNamespaceFilePath only when either is specified
+			// while always updating kernel and fuse mount options.
 			if newCsiClusterConfigEntry.CephFS.SubvolumeGroup != "" || newCsiClusterConfigEntry.CephFS.NetNamespaceFilePath != "" {
 				centry.CephFS = newCsiClusterConfigEntry.CephFS
+			} else {
+				centry.CephFS.KernelMountOptions = newCsiClusterConfigEntry.CephFS.KernelMountOptions
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:194]: Above on line 167 there is a check for `newCsiClusterConfigEntry != nil`, but this usage of the var is not inside that check. Don't we need to check again for nil here?

File: pkg/operator/ceph/csi/cluster_config.go [hunk 4]
```diff
@@ -207,16 +217,9 @@ func updateCsiClusterConfig(curr, clusterKey string, newCsiClusterConfigEntry *C
...
-				centry.RBD = newCsiClusterConfigEntry.RBD
-			}
-			// Add a condition not to fill with empty values
-			if newCsiClusterConfigEntry.CephFS.SubvolumeGroup != "" || newCsiClusterConfigEntry.CephFS.NetNamespaceFilePath != "" {
-				centry.CephFS = newCsiClusterConfigEntry.CephFS
-			}
-			if newCsiClusterConfigEntry.NFS.NetNamespaceFilePath != "" {
-				centry.NFS = newCsiClusterConfigEntry.NFS
-			}
+			centry.RBD = newCsiClusterConfigEntry.RBD
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:220]: Need to check for nil here as well?

File: pkg/operator/ceph/csi/cluster_config.go [hunk 5]
```diff
@@ -170,6 +170,9 @@ func updateCsiClusterConfig(curr, clusterKey string, newCsiClusterConfigEntry *C
 			// update default clusterID's entry
 			if clusterKey == centry.Namespace {
 				centry.Monitors = newCsiClusterConfigEntry.Monitors
+				centry.ReadAffinity = newCsiClusterConfigEntry.ReadAffinity
```
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:173]: What is the difference between the `for` loop on line 168 and the loop lower on line 180? It seems there is a lot of duplication between these two loops.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/cluster_config.go (hunk 3 above)
Comment (Rakshith-R) [reply pkg/operator/ceph/csi/cluster_config.go:194]: > Above on line 167 there is a check for `newCsiClusterConfigEntry != nil`, but this usage of the var is not inside that check. Don't we need to check again for nil here?

The check at line 183 handles the nil case and breaks the for loop before it reaches this line.

**Review by travisn — APPROVED** [review]

**Pull request merged**